### PR TITLE
Delegate toXContent logic from ClusterState to its member classes #48218

### PR DESCRIPTION
**Pull request by zacharymorn**
This is a continuation of PR https://github.com/elastic/elasticsearch/pull/48924, but to approach the problem in the recommended direction,  as that  PR has too many commits to cleanly show the changes.

MetaData.toXContent is already parameterized as it checks for API context for certain serializations. Hence I’ve organized the 4 commits in that PR as follows:

1. Add API and Gateway serialization tests for complex ClusterState and MetaData objects without modifying production code
2. Copy over ClusterState.toXContent serialization test metadata section to  MetaData.toXContent API serialization tests to highlight the difference and setup baseline 
3. Parameterize MetaData.toXContent further and have its serialization under API context to be as close as possible to ClusterState. 
	1. Note that as the updates to existing serialization test cases show, there are inconsistencies of serializations under API context.
	2. We can pick and choose which serializaton scheme we actually want, and which one should no longer be supported under API context.
4. Replace ClusterState.toXContent metadata section with parameterized MetaData.toXContent

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -418,104 +414,15 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-                    if (mapping.size() == 1 && mapping.containsKey(cursor1.key)) {
-                        // the type name is the root value, reduce it
-                        mapping = (Map<String, Object>) mapping.get(cursor1.key);
-                    }
-                    builder.field(cursor1.key);
-                    builder.map(mapping);
-                }
-                builder.endObject();
 
+        Map<String, String> mapParams = new HashMap<>();
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: can you add this as a static field to this class and call it `API_PARAMS`?

File: server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java [hunk 2]
```diff
@@ -246,13 +245,19 @@ public void testSimpleJsonFromAndTo() throws IOException {
         assertThat(indexMetaData.mapping().source().string(), equalTo(MAPPING_SOURCE1));
         assertThat(indexMetaData.getAliases().size(), equalTo(3));
         assertThat(indexMetaData.getAliases().get("alias1").alias(), equalTo("alias1"));
-        assertThat(indexMetaData.getAliases().get("alias1").filter().string(), equalTo(ALIAS_FILTER1));
+        // By default, MetaData.Builder.toXContent(metaData) will use ToXContent.EMPTY_PARAMS, which generates the same serialization
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java]: I think we should remove that static method (MetaData.Builder.toXContent) which seems to be only used by this test.

Instead, we should test both API and non-API serialization in this class, and non-API deserialization.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 3]
```diff
@@ -1318,7 +1360,21 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
                                 throw new IllegalArgumentException("Unexpected token: " + token);
                             }
                         }
-                    } else if ("warmers".equals(currentFieldName)) {
+                    } else if (KEY_PRIMARY_TERMS.equals(currentFieldName)) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: why did you add deserialization code here?

Is it because we are testing API deserialization somewhere? If so, I would rather have the tests changed to test non-API deserialization.

Same comment of other places below.

**Review by ywelsch — COMMENTED** [review]
Thanks for working on this. It's heading in the right direction. I've left some comments (in particular, I'm surprised that we need to change deserialization code).

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 1 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: Sure will do.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 4]
```diff
@@ -1186,48 +1187,89 @@ public IndexMetaData build() {
...
             builder.field(KEY_STATE, indexMetaData.getState().toString().toLowerCase(Locale.ENGLISH));
 
             boolean binary = params.paramAsBoolean("binary", false);
 
             builder.startObject(KEY_SETTINGS);
-            indexMetaData.getSettings().toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", "true")));
+            if (context != MetaData.XContentContext.API) {
+                indexMetaData.getSettings().toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", "true")));
+            }
+            else {
```
[the remark this answers is not in the document]
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Sure will do. If this is style requirement I can add a rule to checkstyle as well?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 5]
```diff
@@ -1186,48 +1187,89 @@ public IndexMetaData build() {
...
+                indexMetaData.getSettings()
+                    .toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", params.param("flat_settings"))));
```
[the remark this answers is not in the document]
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: As before the rest of `params` was not passed along to `IndexMetaData.toXContent`, I suspect this may change the serialization and thus introduce a breaking change. I can try and submit another commit with that to see what happen?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 3 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: The primary reason for adding these is, the original `ClusterState.toXContent` emits array serialization for `aliases` and object serialization for `primary_terms`, which are different from what `MetaData.toXContent` emits, which is object serialization for `aliases` and array serialization for `primary_terms`, under API context. The delta can been seen here https://github.com/elastic/elasticsearch/pull/52743/commits/072e05d370f464ec8582d55c7c26bb8199b58735#diff-4385d650a8cef0592db3454ad42e4bf3L1172-R1157 

In general, as the duplicated code diverges, there are conflicting serializations under API context. 

So the additional deserialization here is to be able to read back in the different serialization emitted by the new code (namely, array serialization for `aliases` and object serialization for `primary_terms`)

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 1 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: Done.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Actually the original `ClusterState#toXContent` uses the whole `params` https://github.com/elastic/elasticsearch/blob/2e617c3b1026bb3fc4063feeb314c1925e83300d/server/src/main/java/org/elasticsearch/cluster/ClusterState.java#L444. I've updated it.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I've put the `else` on the same line for now without adding checkstyle rule.

File: server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java (hunk 2 above)
Comment (zacharymorn) [reply server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java]: I've moved the new serialization tests from `MetaDataTests` to `ToAndFromJsonMetaDataTests`. 

I've inlined / removed 1 `MetaData.Builder#toXContent` which is only used by test. There's another one though that takes more parameters and used in production code, so I left that one alone.

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 6]
```diff
@@ -418,104 +420,13 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-                builder.startObject(IndexMetaData.KEY_IN_SYNC_ALLOCATIONS);
-                for (IntObjectCursor<Set<String>> cursor : indexMetaData.getInSyncAllocationIds()) {
-                    builder.startArray(String.valueOf(cursor.key));
-                    for (String allocationId : cursor.value) {
-                        builder.value(allocationId);
-                    }
-                    builder.endArray();
-                }
-                builder.endObject();
+        API_PARAMS.put("flat_settings", params.param("flat_settings"));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: This mutates the global variable on every call? Certainly not what we want?

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 7]
```diff
@@ -418,104 +420,13 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-            for (ObjectObjectCursor<String, MetaData.Custom> cursor : metaData.customs()) {
-                builder.startObject(cursor.key);
-                cursor.value.toXContent(builder, params);
-                builder.endObject();
-            }
-
-            builder.endObject();
+        // meta data
+        if (metrics.contains(Metric.METADATA)) {
+            metaData.toXContent(builder, new ToXContent.MapParams(API_PARAMS));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: Instead of passing down a limited set of parameters here, I wonder if we should pass all of them, but additionally inject the API context.

Given that Metadata is already using API context as default, I wonder if we need to do anything here in this class. Perhaps it's sufficient to just pass down the params as is?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 3 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I think we should avoid being able to read back the API-serialized XContent. There is no requirement to do so, and it introduces leniency in the parsing here, with the possibility of silently hiding bugs where the wrong context is chosen for serialization.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 8]
```diff
@@ -1270,28 +1270,26 @@ public MetaData build() {
...
-        public static String toXContent(MetaData metaData) throws IOException {
-            XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON);
-            builder.startObject();
-            toXContent(metaData, builder, ToXContent.EMPTY_PARAMS);
-            builder.endObject();
-            return Strings.toString(builder);
-        }
-
         public static void toXContent(MetaData metaData, XContentBuilder builder, ToXContent.Params params) throws IOException {
             XContentContext context = XContentContext.valueOf(params.param(CONTEXT_MODE_PARAM, "API"));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Use `CONTEXT_MODE_API` constant here?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 9]
```diff
@@ -1270,28 +1270,26 @@ public MetaData build() {
...
+                builder.startObject("metadata");
+            }
+            else {
+                builder.startObject("meta-data");
+                builder.field("version", metaData.version());
+            }
 
-            builder.field("version", metaData.version());
             builder.field("cluster_uuid", metaData.clusterUUID);
+            // This is required for some serialization under API context, as some tests in MetaDataTests assert its presenc
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I'm not sure I understand this comment? Is it outdated?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 10]
```diff
@@ -1347,7 +1345,8 @@ public static MetaData fromXContent(XContentParser parser, boolean preserveUnkno
                 currentFieldName = parser.currentName();
             }
 
-            if (!"meta-data".equals(parser.currentName())) {
+            if (!"meta-data".equals(parser.currentName()) &&
+                !"metadata".equals(parser.currentName())) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: As commented above, let's remove the lenient parsing here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 11]
```diff
@@ -1352,7 +1404,13 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
...
                         }
                         builder.primaryTerms(list.toArray());
-                    } else {
+                    }
+                    else if (KEY_ALIASES.equals(currentFieldName)) {
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            builder.putAlias(AliasMetaData.builder(parser.text()).build());
+                        }
+                    }
+                    else {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: move these `else`s up (lot's of other occurrences in this file)

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left more comments. Thanks for iterating with me on this.

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 7 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: That's a good point. The delegation refactoring ultimately do away the need for selecting params in this layer. I've removed it.

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 6 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java]: I've removed `API_PARAMS` as suggested so this is no longer relevant.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 8 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Done.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 9 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Yes sorry. This was added in between the commits, as was meant to say that `ClusterState.toXContent` will now start to emit `cluster_uuid_committed` in its serialization, as I retried to remove `cluster_uuid_committed` before to keep `ClusterState.toXContent` serialization unchanged, but had to put back due to the test asserting it.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 10 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Done. This also caused some test cases to be broken and removed as they relied on deserialization of API-serialized XContent https://github.com/elastic/elasticsearch/pull/52743/commits/0c11613501e979d9202a1f6fa7dda3aca4b69b72

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 11 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I've removed these changes as they are also for reading back API-serialized XContent. I did correct `else` placement in other places though.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 3 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I've given this a shot and removed deserialization for both `primary_terms` and `aliases`. These ended up breaking a lot of tests as they serialized under default context (API), and deserialized for further checking. I thus removed the breaking tests (as little as I can, but more can be removed as what's left are just setup code) and isolated the removal in this commit  https://github.com/elastic/elasticsearch/pull/52743/commits/0c11613501e979d9202a1f6fa7dda3aca4b69b72. 

If we don't want to remove these many tests for such small changes, the other possible approach is to not emit those different serialization for `aliases` and `primary_terms` for `ClusterState.toXContent`. This obviously would break code relying on it.

Please let me know which way you prefer, or if you have other ideas around this.

**Comment by zacharymorn** [comment]
> I've left more comments. Thanks for iterating with me on this.

Sure no problem. I do see these changes to be prone to errors and bugs, and so very appreciate your patience and time for reviewing them carefully!

Once these changes are in good shape, I'll merge in master and resolve conflicts as well.

**Comment by ywelsch** [comment]
@elasticmachine ok to test

**Comment by ywelsch** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 12]
```diff
@@ -1451,7 +1452,13 @@ public static Settings addHumanReadableSettings(Settings settings) {
         return builder.build();
     }
 
-    private static final ToXContent.Params FORMAT_PARAMS = new MapParams(Collections.singletonMap("binary", "true"));
+    private static final ToXContent.Params FORMAT_PARAMS;
```
Comment (zacharymorn) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Didn't realize we could use this to save the tests, good to know!

**Review by ywelsch — APPROVED** [review]
LGTM
